affirm/src/main/java/com/affirm/android/AffirmHttpClient.java

```diff
@@ -0,0 +1,232 @@
+package com.affirm.android;
+
+import com.affirm.android.exception.APIException;
+import com.affirm.android.exception.ConnectionException;
+import com.affirm.android.exception.InvalidRequestException;
+import com.affirm.android.exception.PermissionException;
+import com.affirm.android.model.AffirmError;
+
+import java.io.IOException;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import okhttp3.Call;
+import okhttp3.Headers;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okhttp3.internal.Util;
+import okio.BufferedSink;
+
+import static com.affirm.android.AffirmConstants.X_AFFIRM_REQUEST_ID;
+import static com.affirm.android.AffirmTracker.TrackingEvent.NETWORK_ERROR;
+import static com.affirm.android.AffirmTracker.TrackingLevel.ERROR;
+import static com.affirm.android.AffirmTracker.createTrackingNetworkJsonObj;
+
+final class AffirmHttpClient {
+
+    private OkHttpClient mOkHttpClient;
+
+    private AffirmHttpClient(@Nullable OkHttpClient.Builder builder) {
+        if (builder == null) {
+            builder = new OkHttpClient.Builder();
+        }
+
+        mOkHttpClient = builder.build();
+    }
+
+    static AffirmHttpClient createClient(@Nullable OkHttpClient.Builder builder) {
+        return new AffirmHttpClient(builder);
+    }
+
+    AffirmHttpResponse execute(final AffirmHttpRequest request) throws APIException,
+            PermissionException, InvalidRequestException, ConnectionException {
+        return execute(request, true);
+    }
+
+    AffirmHttpResponse execute(final AffirmHttpRequest request, boolean sendTrackEvent)
+            throws APIException, PermissionException, InvalidRequestException, ConnectionException {
+        Request okHttpRequest = getRequest(request);
+        Call call = mOkHttpClient.newCall(okHttpRequest);
+        try {
+            Response response = call.execute();
+
+            boolean responseSuccess = response.isSuccessful();
+            if (!responseSuccess && sendTrackEvent) {
+                AffirmTracker.track(NETWORK_ERROR, ERROR,
+                        createTrackingNetworkJsonObj(okHttpRequest, response));
+            }
+
+            final Headers headers = response.headers();
+            String requestId = headers.get(X_AFFIRM_REQUEST_ID);
+            if (response.code() < 200 || response.code() >= 300) {
+                ResponseBody responseBody = response.body();
+                final AffirmError affirmError = AffirmPlugins.get().gson()
+                        .fromJson(responseBody != null
```

Comment:
Use fromJson(Reader) not fromJson(String) for performance. You can use response.body().contentLength() == 0L to see if it's empty

Sadly, the servers might return html. For example on 500s. I would be ok with ignoring json AffirmError serializing errors

Reply:
Closing this PR as these issues are now fixed on master branch